File: erlang/seminars/tracer/src/camera.erl
-module(camera).

-export([camera/5, normal/1, size/1, ray/3]).

-record(camera, {pos,    %  the position of the camera
		 corner, % the upper left corner of the plane
		 right,  % the vector in the plane from the corner going right
		 down,   % the vector in the  plane from the corner going down
		 size    % the {width, height}
		}).

camera(Pos, Corner, Right, Down, Size) ->
    #camera{pos = Pos, 
	    corner = Corner,
	    right = Right, 
	    down = Down, 
	    size = Size}.

normal(Size) ->
    {Width, Height} = Size,
    D = Width * 1.2,
    H = Width / 2,
    V = Height /  2,
    Corner = {-H,V,D},
    camera({0,0,0}, Corner, {1,0,0}, {0,-1,0}, Size).


size(#camera{size=Size}) ->
    Size.

ray(X, Y, Camera) ->
    Origin = Camera#camera.pos,
    Xpos = vector:smul(Camera#camera.right, X),
    Ypos = vector:smul(Camera#camera.down, Y),
    Pixle = vector:add(Camera#camera.corner, vector:add(Xpos, Ypos)),
    Dir = vector:normalize(Pixle),
    objects:ray(Origin, Dir).

File: exercises/tracer/src/camera.ex
defmodule Camera do

  defstruct(
    pos: nil,
    corner: nil,
    right: nil,
    down: nil,
    size: nil
  )

  def normal(size) do
    {width, height} = size
    d = width * 1.2
    h = width / 2
    v = height / 2
    pos = {0,0,0}
    corner = {-h, v, d}
    right = {1, 0, 0}
    down = {0, -1, 0}
    %Camera{pos: pos, corner: corner, right: right, down: down, size: size}
  end


  def ray(%Camera{pos: pos, right: right, down: down, corner: corner}, x, y) do
    x_pos = Vector.smul(right, x)
    y_pos = Vector.smul(down, y)
    pixle = Vector.add(corner, Vector.add(x_pos, y_pos))
    dir = Vector.normalize(pixle)
    %Ray{pos: pos, dir: dir}
  end

end
